### ISSUE TITLE
Writing in Ubuntu is extremely slow

### ISSUE THREAD

**Issue by 3x3y3z3t** [tusdotnet/tusdotnet#188]
I made an ASP.NET Core app to manage my self-host storage server, using tusdotnet on server side and tus-js-client on client side.
My problem is: every "writing to disk" operation is extremely slow (after every chunk is uploaded and when the whole file is uploaded). The average write speed is 1.12 MB/s, while `File.Copy()` on the same disk is around 17 MB/s ~ 21 MB/s.

Is there any configuration that I missed that affect writing speed, or that is how tusdotnet work?

**More details:**
- The server runs on Ubuntu Server on a Banana Pi M5 (arm-64 4GB RAM). The app itself runs from an ext4 partition on a WD Red HDD (5400 RPM).
- When a file is completely uploaded on server (`OnFileComplete`), it is written to a directory on the same ext4 partition as the app. Average writing speed here is 1.12 MB/s.
- The file is then copied to another NTFS partition on the same HDD using `File.Copy()`. Average writing speed here is 17 ~ 21 MB/s.
- I measure the time by create a new `DateTime` or `Date` object, perform the copy/upload operation, then subtract it with a new `DateTime` to get the elapsed time.
- After a 64MB chunk is sent, I need to wait ~40s before that chunk is accepted and `onChunkComplete` even is fired.

**tusdotnet config**
- I used the same default config shown on the wiki. Here is how I write the file to disk:
```cs
OnFileCompleteAsync = async (_eventContext) =>
{
    string filename = "...";

    tusdotnet.Interfaces.ITusFile file = await _eventContext.GetFileAsync();
    Dictionary<string, tusdotnet.Models.Metadata> metadata = await file.GetMetadataAsync(_eventContext.CancellationToken);
    using (Stream content = await file.GetContentAsync(_eventContext.CancellationToken)) 
    {
        int bufferSize = 4 * 1024 * 1024;

        FileStream fileStream = File.Create(fileName, bufferSize, FileOptions.Asynchronous);
        await content.CopyToAsync(fileStream, bufferSize);

        await fileStream.DisposeAsync();
    }

    // do additional processing;
}
```

**tus-js-client config**
- I used the same config shown on the tutorial, except I set chunk size to 64 MB due to server limitation.
```js
let uploader = new tus.Upload(filename, {
    // other required data;

    chunkSize: 64 * 1024 * 1024
}
```

**Comment by smatsson** [comment]
Hi,

There seems to be two issues described here, namely slow write speed during upload and slow write speed once your file has been completely uploaded. Let me know if I misunderstood that :) 

**Writing during upload**
By default TusDiskStore will write to disk using a 50 KiB buffer. This was chosen to keep resumability as stable as possible i.e. the client should have to re-upload as little data as possible if a connection breaks down. You can change this buffer size by changing `TusDiskBufferSize` when creating the TusDiskStore instance (https://github.com/tusdotnet/tusdotnet/blob/master/Source/tusdotnet/Stores/TusDiskStore.cs#L76).

On .NET6, the default mode of reading data from the client is through the `System.IO.PipeReader` provided by ASP.NET Core. This is more efficient in a lot of ways but you might end up with slower writes due to how the PipeReader works internally (multiple small buffers instead of one big). If changing the write buffer size does not work, try setting `UsePipelinesIfAvailable` to false and see if the speed picks up (https://github.com/tusdotnet/tusdotnet/blob/master/Source/tusdotnet/Models/DefaultTusConfiguration.cs#L62). Note that this might impact memory usage and CPU load as there are now multiple buffers and copying memory between them.

**Write speed in OnFileComplete**
Doing a `CopyToAsync` will always be slower than doing a `File.Copy` as `CopyToAsync` will read data from one stream, copy it to the other (and possibly write it to disk), read some more, copy it.... while `File.Copy` will just tell the OS to copy the file. Is there a reason here that you do not just use `File.Copy` directly? Or maybe even better a `File.Move` to prevent the copy altogether? 
```csharp
OnFileCompleteAsync = ctx =>
{
    File.Copy(Path.Combine(@"path provided to TusDiskStore", ctx.FileId), "new file path");
    return Task.CompletedTask;
}
```

**Comment by 3x3y3z3t** [comment]
Hi,
Thanks for the detailed explanation. And yes, I got problem with slow write speed during upload and slow write speed once file has been completed.

**About write speed during upload**
I have tried set TusDiskStore buffer size to 8 MiB and the upload speed bumped up to 11 MB/s (~10x faster).

**About write speed in OnFileCompleted**
I haven't had time to try the `File.Copy` approach yet, but setting the buffer size to 8 MiB when creating TusDiskStore seem to be also doing magic here. I noticed from the client that it took significantly less time to finish a file after all chunks have been uploaded.
The reason I didn't go with `File.Copy` in the first place is because I followed the guide in the readme. I thought the only way to retrieve my file is via `ITusFile.GetContentAsync` (as a stream).

---

Now if I set the buffer size to 64 MiB (*equal the size of one chunk*), will I get maximum performance? Will I get diminishing returns because `PipeReader` create more small buffers?
(I don't mind having to re-upload that much data in case of losing connection.)
